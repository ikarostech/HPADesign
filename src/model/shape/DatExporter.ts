import { Vector2 } from "three/src/math/Vector2";

export default function (shape: Vector2[]): string {

  return "";
}