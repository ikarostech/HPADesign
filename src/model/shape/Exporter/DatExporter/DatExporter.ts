import { Vector2 } from "three/src/math/Vector2";

export default function (name: string, shape: Vector2[]): string {
  let content = name + "\n";
  shape.forEach(point => {
    content += " " + point.x.toString() + "     " + point.y.toString() + "\n";
  })
  return content;
}