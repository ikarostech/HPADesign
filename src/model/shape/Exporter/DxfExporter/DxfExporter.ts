import { Vector2 } from "three/src/math/Vector2";

//ファイル名称などのメタデータをファイルに追加する可能性があるためメソッドとして管理
function dxfHeader(): string {
  return "0\nSECTION\n" +
    "2\nHEADER\n" +
    "0\nENDSEC\n" +
    "0\nSECTION\n" +
    "2\nENTITIES\n";
}

function dxfFooter(): string {
  return "0\nENDSEC\n" +
    "0\nEOF\n";
}

//TODO Vector2[]を受け取るのではなくそのラッパーTypeで受け取るようにする
export default function (shape: Vector2[]): string {
  let content = "";
  const polylineHeader = "0\nPOLYLINE\n" +
    "8\n0\n" +
    "66\n1\n" +
    "70\n1\n";
  const polylineFooter = "0\nSEQEND\n" +
    "8\n0\n";

  shape.forEach(point => {
    const vertex: string = "0\nVERTEX\n" +
      "8\n0\n" +
      "10\n" + point.x.toString() + "\n" +
      "20\n" + point.y.toString() + "\n" +
      "30\n0\n";

    content += vertex;
  })
  const polyline: string = polylineHeader + content + polylineFooter;
  return dxfHeader() + polyline + dxfFooter();
}