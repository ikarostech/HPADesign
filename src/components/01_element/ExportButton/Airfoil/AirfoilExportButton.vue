<template>
  <div>
    <v-menu offset-y>
      <template v-slot:activator="{ on }">
        <v-btn color="primary" dark v-on="on">書き出し</v-btn>
      </template>
      <v-list>
        <v-list-item @click="exportDat">
          <v-list-item-title>dat</v-list-item-title>
        </v-list-item>
        <v-list-item @click="exportDxf">
          <v-list-item-title>dxf</v-list-item-title>
        </v-list-item>
      </v-list>
    </v-menu>
  </div>
</template>
<script lang="ts">
import { Vector2 } from "three/src/math/Vector2";
import Vue, { PropType } from "vue";
import DatExporter from "@/model/shape/Exporter/DatExporter/DatExporter";
import DxfExporter from "@/model/shape/Exporter/DxfExporter/DxfExporter";
export default Vue.extend({
  props: {
    shape: Array as PropType<Vector2[]>,
  },
  methods: {
    exportDat(): void {
      const datText = DatExporter("airfoil", this.shape);
      this.downloadContent("airfoil.dat", datText);
    },
    exportDxf(): void {
      const dxfText = DxfExporter(this.shape);
      this.downloadContent("airfoil.dxf", dxfText);
    },
    downloadContent(filename: string, content: string): void {
      const a = document.createElement("a");
      a.href = "data:text/plane," + encodeURIComponent(content);
      a.download = filename;
      a.style.display = "none";
      document.body.appendChild(a);
      a.click();
      document.body.removeChild(a);
    },
  },
});
</script>