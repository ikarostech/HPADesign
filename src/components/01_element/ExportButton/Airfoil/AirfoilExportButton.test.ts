import DatExporter from '@/model/shape/Exporter/DatExporter/DatExporter';
import DxfExporter from '@/model/shape/Exporter/DxfExporter/DxfExporter';
import { any } from '@tensorflow/tfjs';
import { shallowMount } from '@vue/test-utils';
import { Vector2 } from 'three/src/math/Vector2';
import AirfoilExportButton from './AirfoilExportButton.vue';

describe('AirfoilExportButton', () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallowMount(AirfoilExportButton, {
      propsData: {
        shape: [new Vector2(1, 1), new Vector2(2, 2)],
      }
    });
  })
  afterEach(() => {
    wrapper.destroy();
    jest.resetAllMocks();
  })
  it('snapshot test', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it('datエクスポートがクリックされるとexportDatが走る', async () => {
    const datElement = wrapper.findAll("v-list-item").at(0);
    expect(datElement.text()).toEqual("dat");

    wrapper.setMethods({
      exportDat: jest.fn()
    });
    await datElement.trigger("click")

    expect((wrapper.vm as any).exportDat).toHaveBeenCalled();
  });

  it('dxfエクスポートがクリックされるとexportDxfが走る', async () => {
    const dxfElement = wrapper.findAll("v-list-item").at(1);
    expect(dxfElement.text()).toEqual("dxf");

    wrapper.setMethods({
      exportDxf: jest.fn()
    });
    await dxfElement.trigger("click")

    expect((wrapper.vm as any).exportDxf).toHaveBeenCalled();
  });

  it('exportDatを実行するとDatExporterとdownloadContentを実行する', () => {

    (DatExporter as any) = jest.fn()

    wrapper.setMethods({
      downloadContent: jest.fn()
    });

    (wrapper.vm as any).exportDat();
    expect(DatExporter).toHaveBeenCalled();
    expect((wrapper.vm as any).downloadContent).toHaveBeenCalled();
  });

  it('exportDatを実行するとDxfExporterとdownloadContentを実行する', () => {
    (DxfExporter as any) = jest.fn()

    wrapper.setMethods({
      downloadContent: jest.fn()
    });
    (wrapper.vm as any).exportDat();

    expect((wrapper.vm as any).downloadContent).toHaveBeenCalled();
  });
})