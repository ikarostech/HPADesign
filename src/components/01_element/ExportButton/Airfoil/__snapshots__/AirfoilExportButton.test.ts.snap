// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirfoilExportButton snapshot test 1`] = `
<div>
  <v-menu
    offset-y=""
  >
     
    <v-list>
      <v-list-item>
        <v-list-item-title>
          dat
        </v-list-item-title>
      </v-list-item>
       
      <v-list-item>
        <v-list-item-title>
          dxf
        </v-list-item-title>
      </v-list-item>
    </v-list>
  </v-menu>
</div>
`;
